### PR TITLE
[SPARK-7407][MLLIB] use uid + name to identify parameters

### PR DESCRIPTION
**Pull request by mengxr**
A param instance is strongly attached to an parent in the current implementation. So if we make a copy of an estimator or a transformer in pipelines and other meta-algorithms, it becomes error-prone to copy the params to the copied instances. In this PR, a param is identified by its parent's UID and the param name. So it becomes loosely attached to its parent and all its derivatives. The UID is preserved during copying or fitting. All components now have a default constructor and a constructor that takes a UID as input. I keep the constructors for Param in this PR to reduce the amount of diff and moved `parent` as a mutable field.

This PR still needs some clean-ups, and there are several spark.ml PRs pending. I'll try to get them merged first and then update this PR.

@jkbradley

### CODE REVIEW

**Comment by jkbradley** [comment]
@mengxr Do we really need to have Param store a string for a parent instead of the actual parent?  I rather liked being able to say "lr.maxIter" and have it print out a full explanation.  I figured we could convert the parent to a string when the Param needed to be applied.

**Comment by jkbradley** [comment]
Btw, is this still a WIP, and if so, what else do you plan to put into this PR?

**Comment by mengxr** [comment]
@jkbradley If a Param is identified by the parents' uid and its name, `Param.toString` shouldn't output the default value and user-supplied value for a particular parent, since a param is not attached to a single parent but also its derivatives. I think this is all I want to add into this PR. I was waiting for the OneVsRest to go in first and update this.

**Comment by jkbradley** [comment]
Right, that makes sense.  OK, I'll check the OneVsRest PR now.

File: examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.examples.ml
 
+import org.apache.spark.ml.util.Identifiable
```
Comment (jkbradley) [inline examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala]: Organize imports

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 2]
```diff
@@ -171,7 +174,7 @@ class Pipeline extends Estimator[PipelineModel] {
  */
 @AlphaComponent
 class PipelineModel private[ml] (
-    override val parent: Pipeline,
+    val uid: String,
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: override?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 3]
```diff
@@ -54,7 +54,7 @@ final class Bucketizer private[ml] (override val parent: Estimator[Bucketizer])
       "bucket, which also includes y. The splits should be strictly increasing. " +
       "Values at -inf, inf must be explicitly provided to cover all Double values; " +
       "otherwise, values outside the splits specified will be treated as errors.",
-    Bucketizer.checkSplits)
+    Bucketizer.checkSplits _)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Curious: Why is this needed?

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 4]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.ml.feature
 
 import org.apache.spark.SparkException
 import org.apache.spark.annotation.AlphaComponent
+import org.apache.spark.ml.util.Identifiable
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: organize imports

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 5]
```diff
@@ -27,6 +27,8 @@ import scala.collection.JavaConverters._
 import org.apache.spark.annotation.AlphaComponent
 import org.apache.spark.ml.util.Identifiable
 
+import scala.util.Try
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: organize imports

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 6]
```diff
@@ -379,7 +427,6 @@ trait Params extends Identifiable with Serializable {
    * @param value  the default value
    */
   protected final def setDefault[T](param: Param[T], value: T): this.type = {
-    shouldOwn(param)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Why was this removed?  Shouldn't it check for ownership?

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 7]
```diff
@@ -303,7 +306,9 @@ class ALS extends Estimator[ALSModel] with ALSParams {
       maxIter = $(maxIter), regParam = $(regParam), implicitPrefs = $(implicitPrefs),
       alpha = $(alpha), nonnegative = $(nonnegative),
       checkpointInterval = $(checkpointInterval), seed = $(seed))
-    copyValues(new ALSModel(this, $(rank), userFactors, itemFactors))
+    val model = new ALSModel(uid, $(rank), userFactors, itemFactors)
+      .setParent(this)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: fits on 1 line

File: mllib/src/main/scala/org/apache/spark/ml/util/Identifiable.scala [hunk 8]
```diff
@@ -19,15 +19,24 @@ package org.apache.spark.ml.util
...
+
 /**
- * Object with a unique id.
+ * Trait for an object with an immutable unique ID that identifies itself and its derivatives.
  */
-private[ml] trait Identifiable extends Serializable {
+trait Identifiable {
+
+  /**
+   * An immutable unique ID for the object and its derivatives.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/Identifiable.scala]: How user-facing is this?  Should we provide more info/promises about how it is used, or should we wait until that is clearer to us?

File: mllib/src/test/java/org/apache/spark/ml/util/IdentifiableSuite.scala [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  test("Identifiable") {
+    val test0 = new Test("test_0")
+    assert(test0.uid === "test_0")
+
+    val test1 = new Test
+    assert(test1.uid.startsWith("test_"))
+  }
+}
+
+object IdentifiableSuite {
```
Comment (jkbradley) [inline mllib/src/test/java/org/apache/spark/ml/util/IdentifiableSuite.scala]: make private?

**Comment by jkbradley** [comment]
@mengxr A bunch of small comments, but other than those, it looks good.  At some point, it might be nice to add a test suite helper method which does standard checks to make sure models are handled correctly.  E.g.: take (Estimator, Model, paramMap passed to fit()), and check UID, parent, and parameter values copied to the model.  I'll make a JIRA for that.

File: examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala (hunk 1 above)
Comment (mengxr) [inline examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala (hunk 2 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala (hunk 3 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Actually I don't know. Both Scala and Intellij complain if I don't put the underscore to make it anonymous.

~~~
[error] /Users/meng/src/spark/mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala:57: missing arguments for method checkSplits in object Bucketizer;
[error] follow this method with `_' if you want to treat it as a partially applied function
[error]     Bucketizer.checkSplits)
~~~

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala (hunk 4 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala (hunk 5 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala (hunk 6 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: This causes some racing condition. Because not all members are initialized before `setDefault` is called. `shouldOwn` checks `params`, which is a lazy val. If `params` is initialized before some params, it would contains `null`. I didn't figure out a good solution. Given that this is not called by users, I removed this check and left a note on `params`.

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala (hunk 7 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/util/Identifiable.scala (hunk 8 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/util/Identifiable.scala]: Good question! I don't think the `Identifiable` interface is user facing, not developer facing either. The only useful function to developers is the `randomUID` under `object Identifiable`.

File: mllib/src/test/java/org/apache/spark/ml/util/IdentifiableSuite.scala (hunk 9 above)
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/util/IdentifiableSuite.scala]: This is in the test suite.

**Comment by mengxr** [comment]
Merged into master and branch-1.4. I will send another PR for SPARK-7609.
